File: A-star.cpp
#include <iostream>
#include <math.h>
#include <ctime>
#include <cstdlib>
#include <conio.h>
#include <vector>
#include <algorithm>
#include <Windows.h>

using namespace std;

class Cell;
float distance(Cell start, Cell finish);

const int HEIGHT = 20, WIDTH = 60;
const float D = 1, D2 = 1.4;

// MAP
// # - let
// . - free cell

/* BIG MAP 40x100
const char map[HEIGHT][WIDTH] = {
	"...############################################################################################....",
	"...#...............................#..........................#...#......#....................#....",
	"...################### #############..........#...............#...#....................#......#....",
	"...#...............................################ #######...#####################.###########....",
	"...#..............#................#.............#............#......#...............#........#....",
	"..................#................#.............#............#......................#........#....",
	"...#..............##################..........................#......#........................#....",
	"...#..............#......#.........#.............#............#############.###################....",
	"...####### ########......#.......................#.############.........#.....................#....",
	"...#.....#.....#.........#.........###############............#.........#............##########....",
	"...#...........#.........#.........#.............#............#.........#..............#...........",
	"...### ##### ###.........#.........#.............#............######.############......#......#....",
	"...#...........#.........#.############## #################.###........#........#......#......#....",
	"...#.......................#................#.................#........#......................#....",
	"...#...........#...........#................#.......#..................#........#......#####.##....",
	"...################### #####................#########.........#.................#.............#....",
	"...#......#................######### #........................#######################.#########....",
	"...#......#..........#####.#.........#### ###.................#.....................#.........#....",
	"...#......#..........#.....#.......#.#......#.................#...........#.........#.........#...",
	"...#......#..........#.....#.......#.#......#.................#...........#.........#.........#....",
	"...#......##########...............#..........#...............#...........###########.........#....",
	"...#.................#.....#.......#..........................#...........#.........#.........#....",
	"...################### #############..........#...............#...........#...................#....",
	"...#...............................################ #######...#.....................#.........#....",
	"...#..............#................#............#.............##########################.######....",
	"..................#................#............#...........#...........#...........#.........#....",
	"...#..............###########################.###.............#.....................#.........#....",
	"...#..............#......#.........#............#.#############.........#.....................#....",
	"...####### ########......#......................#.......#.....#.........#...........#.........#....",
	"...#.....#.....#.........#.........#....................#.....#.........#...........#.........#....",
	"...#...........#.........#.........#............#.............###########...........#.........#....",
	"...### ##### ###.........#.........#............#.......#.....#.....................#.........#....",
	"...#...........#.........#.############## ###################.#.........#...........#.........#....",
	"...#.......................#................#........#........#.........#...........######.####....",
	"...#...........#...........#................#........#........#.........#...........#.........#....",
	"...################### #####................#######.##........#.........#...........#.........#....",
	"...#......#................######### #...............#........#.........#...........#.........#....",
	"...#.................#####.#.........#### ###........#........#.........#.....................#....",
	"...#......#..........#.....#.........#......#.................#.........#...........#.........#....",
	"...############################################################################################...."
};
*/

/*
const char map[HEIGHT][WIDTH] = {
	"......#.....",
	"......#.....",
	"............",
	"............",
	"............"
};
*/

const char map[HEIGHT][WIDTH] = {
	"...#...............................#..........#............",
	"...#...............................#.......................",
	"...################### #############..........#............",
	"...#...............................################ #######",
	"...#..............#................#.......................",
	"..................#................#.......................",
	"...#..............##################.......................",
	"...#..............#................#.......................",
	"...####### ##############.##...............................",
	"...#...........#...........#.#######.......................",
	"...#...........#...........#.......#.......................",
	"...#...........#...........#.......#.......................",
	"...#...........#...........############## #################",
	"...#.......................#................#..............",
	"...#...........#...........#................#..............",
	"...################### #####................#..............",
	"...#......#................######### #.....................",
	"...#......#..........#####.#.........#### ###..............",
	"...#.................#.....#.........#......#..............",
	"...#......#..........#.....#.........#......#.............."
};

// Colors codes
enum ConsoleColor
{
	Black = 0,
	Blue = 1,
	Green = 2,
	Cyan = 3,
	Red = 4,
	Magenta = 5,
	Brown = 6,
	LightGray = 7,
	DarkGray = 8,
	LightBlue = 9,
	LightGreen = 10,
	LightCyan = 11,
	LightRed = 12,
	LightMagenta = 13,
	Yellow = 14,
	White = 15
};

// Cursor Position
void gotoxy(int x, int y) {
	COORD coord = { x, y };
	SetConsoleCursorPosition(GetStdHandle(STD_OUTPUT_HANDLE), coord);
}

// Seting Cursor Color
void SetColor(int text, int background)
{
	HANDLE hStdOut = GetStdHandle(STD_OUTPUT_HANDLE);
	SetConsoleTextAttribute(hStdOut, (WORD)((background << 4) | text));
}

class Cell {
public:
	int x, y;
	char sym;
	bool isLet;
	int cameFrom;
	float f; // Sum of g and h
	float g; // Cost of step
	float h; // Heuristic (Aproximate distance to goal)

	void calcF() { f = g + h; }
	Cell() {};

	Cell(int X, int Y, bool state = true) :x(X), y(Y), isLet(state) { };

	void setPos(int X, int Y) {
		x = X;
		y = Y;
	};

	bool operator==(Cell obj) {
		return this->x == obj.x && this->y == obj.y;
	}

	// Returns neighbors
	vector <Cell> getNeighbors(Cell area[][WIDTH]) {
		vector <Cell> neighbors;

		for (int i = 0; i < HEIGHT; i++)
			for (int j = 0; j < WIDTH; j++) {
				if (abs(area[i][j].x - this->x) <= 1 && abs(area[i][j].y - this->y) <= 1 && !(area[i][j] == *this))
				{
					neighbors.push_back(area[i][j]);
				}
			}
		return neighbors;
	}
};

// Heuristic Euclidean function
/*float distance(Cell start, Cell finish) {
	return sqrt(pow(finish.x - start.x, 2) + pow(finish.y - start.y, 2));
}*/

// Heuristic Manhattan function 
float distance(Cell start, Cell finish) {
	float dx = abs(start.x - finish.x);
	float dy = abs(start.y - finish.y);
	return D * (dx + dy) + (D2 - 2 * D) * min(dx, dy);
}

int main()
{
	system("cls");
	srand(clock());
	int i, j;

	Cell area[HEIGHT][WIDTH];

	for (i = 0; i < HEIGHT; i++) {
		for (j = 0; j < WIDTH; j++) {
			area[i][j].setPos(j, i);
			area[i][j].g = (float)INT_MAX;
			area[i][j].f = (float)INT_MAX;
			if (map[i][j] == '#') {
				area[i][j].isLet = false;
				area[i][j].sym = '#';
			}
			else {
				area[i][j].isLet = true;
				area[i][j].sym = '.';
			}
		}
	}

	vector <Cell> openSet;

	int startX = 0, startY = 1;
	/*do {
		startX = rand() % WIDTH;
		startY = rand() % HEIGHT;
	} while (area[startY][startX].isLet == false);
	*/
	int targetX = 43, targetY = 1;
	/*do {
		targetX = rand() % WIDTH;
		targetY = rand() % HEIGHT;
	} while (area[targetY][targetX].isLet == false);
	*/

	openSet.push_back(Cell(startX, startY));
	openSet[0].g = 0;
	openSet[0].f = distance(openSet[0], area[targetY][targetX]);
	openSet[0].cameFrom = -1;

	vector <Cell> path; // Path Cells
	int k = 0; // Path cell index

	clock_t start = clock();

	cout << "Finding best path..." << endl;

	while (openSet.size() != 0) {
		Cell current;
		int minF = 0;

		for (i = 0; i < openSet.size(); i++)
			if (openSet[minF].f > openSet[i].f) minF = i;

		current = openSet[minF];
		path.push_back(current);

		// cout << "Current Pos. X: " << current.x << " Y: " << current.y << endl;

		if (current == area[targetY][targetX]) {
			cout << "Path found!" << endl;
			break;
		}
		// Removing current cell
		vector <Cell>::iterator iter = find(openSet.begin(), openSet.end(), current);
		openSet.erase(iter);

		vector <Cell> neighbors = current.getNeighbors(area);

		for (i = 0; i < neighbors.size(); i++)
		{
			float g_tmp;
			if(abs(current.x - neighbors[i].x) == 1 && abs(current.y - neighbors[i].y) == 1)
				g_tmp = current.g + D2;
			else
				g_tmp = current.g + D;

			if (g_tmp < neighbors[i].g && neighbors[i].isLet == true) {

				neighbors[i].cameFrom = k;
				neighbors[i].g = g_tmp;
				neighbors[i].h = ceill(distance(neighbors[i], area[targetY][targetX]));
				// neighbors[i].h = distance(neighbors[i], area[targetY][targetX]);
				neighbors[i].calcF();

				if (find(openSet.begin(), openSet.end(), neighbors[i]) == openSet.end())
					openSet.push_back(neighbors[i]);
			}
		}
		k++;
	}
	Cell current = openSet[openSet.size() - 1];
	do {
		current = path[current.cameFrom];
		area[current.y][current.x].sym = '+';
	} while (current.cameFrom != -1);

	area[startY][startX].sym = 'S';
	area[targetY][targetX].sym = 'F';

	cout << endl;
	// Drawing area
	for (i = 0; i < HEIGHT; i++) {
		for (j = 0; j < WIDTH; j++) {
			if (area[i][j].sym == '+') SetColor(Yellow, Black);
			if (area[i][j].sym == 'S') SetColor(LightGreen, Black);
			if (area[i][j].sym == 'F') SetColor(LightRed, Black);
			if (area[i][j].sym == '.') SetColor(DarkGray, Black);
			if (area[i][j].sym == '#') SetColor(Red, Black);
			cout << area[i][j].sym;
		}
		cout << endl;
	}

	SetColor(7, 0);
	char ans;
	cout << "Start (X: " << startX << "; Y: " << startY << ")" << endl;
	cout << "End (X: " << targetX << "; Y: " << targetY << ")" << endl;
	cout << "Path finded in " << clock() - start << "ms" << endl << endl;
	do {
		cout << "Try again? (Y/n): ";
		ans = (char)_getch();

		if (ans == 'y' || ans == 'Y') main();
		if (ans == 'n' || ans == 'N') exit(0);
		cout << endl;
	} while (true);

	return 0;
}

File: A-star.py
from colorama import init
from colorama import Fore, Back, Style

import time;
import os;
import random;
import math;

init()

D = 1;
D2 = 1.4;

HEIGHT = 20;
WIDTH = 59;

map = [
	"...#...............................#..........#............",
	"...#...............................#.......................",
	"...##.##############################..........#............",
	"...#...............................#######.################",
	"...#..............#................#.......................",
	"..................#................#.......................",
	"...#..............##################.......................",
	"...#..............#................#.......................",
	"...####.#################.##...............................",
	"...#...........#...........#.#######.......................",
	"...#...........#...................#.......................",
	"...#...........#...........#.......#.......................",
	"...#...........#...........################################",
	"...#........................................#..............",
	"...#...........#...........#................#..............",
	"...####.##########.#########................#..............",
	"...#......#................###########.....................",
	"...#......#..........#####.#.........#### ###..............",
	"...#.................#.....#.........#......#..............",
	"...#......#..........#.....#.........#......#.............."
];

# Cell class
class Cell:
	sym = ''
	cameFrom = -1;
	isLet = None;
	f = float('inf');
	g = float('inf');

	def __init__ (self, x, y):
		self.x = x;
		self.y = y;

	def getNeighbors (self, area):
		neighbors = [];

		for i in range(0, HEIGHT):
			for j in range(0, WIDTH):
				if abs(area[i][j].x - self.x) <= 1 and abs(area[i][j].y - self.y) <=1 and area[i][j] != self:
					neighbors.append(area[i][j]);

		return neighbors;


# Manhattan heruistic function
def heruistic (start, finish):
	dx = abs(start.x - finish.x);
	dy = abs(start.y - finish.y);

	return D * (dx + dy) + (D2 - 2 * D) * min(dx, dy);


def aStar (start, target, area):
	path = [];
	openSet = [];

	for i in range(0, HEIGHT):
		area.append([]);
		for j in range(0, WIDTH):
			# print(i , ' ' , j);
			area[i].append( Cell(j ,i) );

			if map[i][j] == '#':
				area[i][j].isLet = True;
				area[i][j].sym = '#';
			else:
				area[i][j].isLet = False;
				area[i][j].sym = '.';

	openSet.append(start);
	openSet[0].g = 0;
	openSet[0].f = heruistic(openSet[0], target);

	while len(openSet) != 0:
		current = openSet[0];

		for x in openSet:
			if current.f > x.f:
				current = x;

		if current.x == target.x and current.y == target.y:
			print("Path founded!");
			current = openSet [ len(openSet) - 1].cameFrom;

			while current.cameFrom != -1:
				#print(current.x,  " ", current.y);
				path.append(current);
				current = current.cameFrom;

			return path;

		openSet.remove(current);

		neighbors = current.getNeighbors(area);
		#print(len(neighbors));

		for neighbor in neighbors:
			
			gTemp = current.g + D;

			if gTemp < neighbor.g and neighbor.isLet == False:
				neighbor.cameFrom = current;
				neighbor.g = gTemp;
				neighbor.f = math.ceil( gTemp + heruistic(neighbor, target) );

				if openSet.count(neighbor) == 0:
					openSet.append( neighbor );

	print("Path not found!");

os.system('cls');
area = [];

while True:
	startX = random.randint(0, WIDTH - 1);
	startY = random.randint(0, HEIGHT - 1);

	if map[startY][startX] == '.':
		break;
start = Cell(startX, startY);

while True:
	targetX = random.randint(0, WIDTH - 1);
	targetY = random.randint(0, HEIGHT - 1);

	if map[targetY][targetX] == '.':
		break;

target = Cell(targetX, targetY);

startTime = int(round(time.time() * 1000))
path = aStar(start, target, area);

if path != None:
	endTime = int(round(time.time() * 1000)) - startTime;

	for cell in path:
		area[cell.y][cell.x].sym = '+';

	area[start.y][start.x].sym = 'S';
	area[target.y][target.x].sym = 'F';

	for i in range(0, HEIGHT):
		for j in range(0, WIDTH):
			if area[i][j].sym == '+':
				print(Fore.YELLOW + Style.BRIGHT, end='');
			if area[i][j].sym == '#':
				print(Fore.RED + Style.NORMAL, end='');
			if area[i][j].sym == '.':
				print(Fore.BLACK + Style.BRIGHT, end='');
			if area[i][j].sym == 'S':
				print(Fore.GREEN + Style.BRIGHT, end='');
			if area[i][j].sym == 'F':
				print(Fore.RED + Style.BRIGHT, end='');
			print(area[i][j].sym, end='');

			print(Fore.RESET + Style.RESET_ALL, end='')
		print('\n', end='');

	print('Time elasped:', endTime, 'ms');
else:
	print('Start (', startX, ',', startY, ')');
	print('Target (', targetX, ',', targetY, ')');